libraries/WiFi/src/WiFiClient.h

```diff
@@ -95,6 +97,8 @@ class WiFiClient : public ESPLwIPClient
 
     IPAddress remoteIP() const;
     IPAddress remoteIP(int fd) const;
+    IPv6Address remoteIP6() const;
+    IPv6Address remoteIP6(int fd) const;    
     uint16_t remotePort() const;
     uint16_t remotePort(int fd) const;
     IPAddress localIP() const;
```

Comment:
maybe we can just add globalIP6 and localIP6? This will keep it simple for Arduino users. We can iterate and by prefix identify who is local and who is global.
Privacy enhancements as far as i know is not available in lwip, yet.

Reply:
Be careful with terminology not to mix up the concepts of "local" and "link-local", as they are different things.

Local vs remote: Local means this end of the connection, i.e. the IP address of the device, as opposed to remote, which is the other end.

Link-local vs global are scopes: Link-local is confined to the local link only, global is a routeable.

You can have both.

With IPv4 you usually only have one *local* address, and it usually has a *global* scope. With IPv6 you have multiple local addresses, both link-local and global scopes.

Generally the global scoped addresses are more useful, e.g. to use as a server address to connect to a device, or when looking in logs. (Although not always, e.g. if the destination machines is on the same link, then IPv6 will use the link-local address.)

Reply:
Examples, from a SIMCOM GSM NB-IoT library I am working on: https://github.com/sgryphon/AdvancedGsmClient/blob/main/src/SIM7020/SIM7020GsmModem.cpp#L470

When I connect the NB-IoT modem, it gets assigned 3 local addresses, plus it's address list includes a localhost address:

Local addresses:
1. 2001:8004:4810:0:719d:1439:899a:42d7 - local IPv6 with global scope
2. 10.88.134.167 - local IPv4 with global scope
3. fe80:0:0:0:719d:1439:899a:42d7 - local IPv6 with link-local scope
4. 127.0.0.1  - local IPv4 with link-local scope (loopback addresses count as link-local).

The actual order the addresses are reported by the device is fe80:, 2001:, 10.88, 127.0, however my `getLocalIPs()` method sorts them from largest scope (global) to smallest scope (link-local), and then within each scope according to the RFC 6724 precedence rules (which includes IPv6 before IPv4), to get the order listed above.

Reply:
For my local desktop computer, sorting my addresses a similar way (but using the bytes; my simple GSM algorithm is using strings and so sorts ':' differently), to sort them in order of usefulness:

1. 2407:8800:bc61:1300::8e6 - local IPv6 with global scope
2. 2407:8800:bc61:1300:4a23:22d5:6c48:f0e8 - local IPv6 with global scope
3. 2407:8800:bc61:1300:4df2:1392:9dd7:d2d7 - local IPv6 with global scope
4. 2407:8800:bc61:1300:ee48:5c84:ad87:c65c - local IPv6 with global scope
5. 192.168.1.195 - local IPv4 with global scope
6. fd7c:e25e:67e8::8e6 - local IPv6 with global scope, in ULA range
7. fd7c:e25e:67e8:0:2c73:36d3:d325:93df - local IPv6 with global scope, in ULA range
8. fd7c:e25e:67e8:0:9ac7:63c9:5f26:b026 - local IPv6 with global scope, in ULA range
9. fd7c:e25e:67e8:0:a9ed:c900:59fa:5449 - local IPv6 with global scope, in ULA range
10. fe80::5f65:961a:14b4:6263 - local IPv6 with link-local scope

Global scope is sorted before link-local (because it is more useful to contact me as a server), with the precedence order IPv6, then IPv4, then IPv6 unique local address (ULA).

As a server the ::8e6 address is better, although outgoing logs will use the temporary addresses (probably not an issue for Arduino). If I was communicating with an IPv4 only server, then it would see the IPv4 address.

So, to really investigate I would need to list all of my addresses (e.g. in logs). Even the ULA addresses could be used, for local communications, and the IPv6 link-local address could even be used in some cases.

Reply:
If you really want to stick with separate properties for IPv6 and IPv4, for only one of each type, then you want to try and return a global scope address if possible:
* `localIP4()` - Shows the local IPv4 address (this will usually be a global scope address); there is usually only one. To be consistent, in case there is more than one IPv4, you should also sort by scope (e.g. if a device has two IP addresses 169.254.1.2 and 10.1.2.3, then show 10.1.2.3 ... although that would be really weird as IPv4 doesn't work that way).
* `localIP6()` - shows the first IPv6 address, sorting global scope before link-local scope.
* `localIP()` - for a single address should show the biggest scope, highest precedence address. i.e. global IPv6 if you have one, otherwise global IPv4, and then finally (if you have no global addresses), then the link-local IPv6.

I do think you should have a getLocalIPs() that returns all of the addresses, which you could then use to select out the other values, e.g. for logging, although in the case of the GSM code I know it only ever gets max 4 addresses (i.e. no temporary, only one IPv6 prefix, etc), so I just log the first 2 address, which will be the one global IPv6 and the one global IPv4.